flink-runtime/src/test/java/org/apache/flink/runtime/checkpoint/channel/SequentialChannelStateReaderImplTest.java

```diff
@@ -0,0 +1,321 @@
+import org.apache.flink.core.memory.MemorySegmentFactory;
+import org.apache.flink.runtime.checkpoint.OperatorSubtaskState;
+import org.apache.flink.runtime.checkpoint.StateObjectCollection;
+import org.apache.flink.runtime.checkpoint.TaskStateSnapshot;
+import org.apache.flink.runtime.io.network.buffer.Buffer;
+import org.apache.flink.runtime.io.network.buffer.FreeingBufferRecycler;
+import org.apache.flink.runtime.io.network.buffer.NetworkBuffer;
+import org.apache.flink.runtime.io.network.buffer.NetworkBufferPool;
+import org.apache.flink.runtime.io.network.partition.ResultPartition;
+import org.apache.flink.runtime.io.network.partition.ResultPartitionBuilder;
+import org.apache.flink.runtime.io.network.partition.ResultSubpartition.BufferAndBacklog;
+import org.apache.flink.runtime.io.network.partition.ResultSubpartitionView;
+import org.apache.flink.runtime.io.network.partition.consumer.BufferOrEvent;
+import org.apache.flink.runtime.io.network.partition.consumer.InputGate;
+import org.apache.flink.runtime.io.network.partition.consumer.SingleInputGate;
+import org.apache.flink.runtime.io.network.partition.consumer.SingleInputGateBuilder;
+import org.apache.flink.runtime.jobgraph.OperatorID;
+import org.apache.flink.runtime.state.InputChannelStateHandle;
+import org.apache.flink.runtime.state.ResultSubpartitionStateHandle;
+import org.apache.flink.runtime.state.memory.ByteStreamStateHandle;
+import org.apache.flink.util.function.ThrowingConsumer;
+
+import org.junit.Test;
+import org.junit.runner.RunWith;
+import org.junit.runners.Parameterized;
+
+import java.io.ByteArrayOutputStream;
+import java.io.DataOutputStream;
+import java.io.IOException;
+import java.util.ArrayList;
+import java.util.Collections;
+import java.util.HashMap;
+import java.util.List;
+import java.util.Map;
+import java.util.Optional;
+import java.util.Random;
+import java.util.function.BiFunction;
+import java.util.function.Function;
+
+import static java.util.function.Function.identity;
+import static java.util.stream.Collectors.toList;
+import static java.util.stream.Collectors.toMap;
+import static java.util.stream.IntStream.range;
+import static org.junit.Assert.assertEquals;
+import static org.junit.Assert.assertTrue;
+
+/**
+ * {@link SequentialChannelStateReaderImpl} Test.
+ */
+@RunWith(Parameterized.class)
+public class SequentialChannelStateReaderImplTest {
+
+	@Parameterized.Parameters(name = "{0}: stateParLevel={1}, statePartsPerChannel={1}, stateBytesPerPart={2},  parLevel={4}, bufferSize={5}")
+	public static Object[][] parameters() {
+		return new Object[][]{
+			{"NoStateAndNoChannels", 0, 0, 0, 0, 0},
+			{"NoState", 0, 10, 10, 10, 10},
+			{"ReadPermutedStateWithEqualBuffer", 10, 10, 10, 10, 10},
+			{"ReadPermutedStateWithReducedBuffer", 10, 10, 10, 20, 10},
+			{"ReadPermutedStateWithIncreasedBuffer", 10, 10, 10, 10, 20},
+		};
+	}
+
+	private final ChannelStateSerializer serializer;
+	private final Random random;
```

Comment:
I think this wouldn't give performance benefit as tests are currently run sequentially.
And even if run in parallel, a new class instance would be created for each test, so threads wouldn't share `Random` instance.

Reply:
The issue is rather that `Random` is always strictly slower than `ThreadLocalRandom` because of lock acquisitions. However, you are right, as long as we not using multiple threads, we do not have contention, so the difference is marginal.

Reply:
> Random is always strictly slower than ThreadLocalRandom

`nextBytes`  used in this test is not overridden in `ThreadLocalRandom` so it's the same.
And the creation of `Random` seems cheaper than `ThreadLocalRandom`.
So it would be slower in the end.